subsys/llext/llext.c

```diff
@@ -0,0 +1,470 @@
+
+	elf_shdr_t shdr;
+	size_t pos = ldr->hdr.e_shoff;
+	unsigned int str_cnt = 0;
+
+	ldr->sect_map = k_heap_alloc(&llext_heap, ldr->hdr.e_shnum * sizeof(uint32_t), K_NO_WAIT);
+	if (!ldr->sect_map) {
+		return -ENOMEM;
+	}
+	ldr->sect_cnt = ldr->hdr.e_shnum;
+
+	ldr->sects[LLEXT_SECT_SHSTRTAB] =
+		ldr->sects[LLEXT_SECT_STRTAB] =
+		ldr->sects[LLEXT_SECT_SYMTAB] = (elf_shdr_t){0};
+
+	/* Find symbol and string tables */
+	for (i = 0; i < ldr->hdr.e_shnum && str_cnt < 3; i++) {
+		llext_seek(ldr, pos);
+		llext_read(ldr, &shdr, sizeof(elf_shdr_t));
+
+		pos += ldr->hdr.e_shentsize;
+
+		LOG_DBG("section %d at %x: name %d, type %d, flags %x, addr %x, size %d",
+			i,
+			ldr->hdr.e_shoff + i * ldr->hdr.e_shentsize,
+			shdr.sh_name,
+			shdr.sh_type,
+			shdr.sh_flags,
+			shdr.sh_addr,
+			shdr.sh_size);
+
+		switch (shdr.sh_type) {
+		case SHT_SYMTAB:
+		case SHT_DYNSYM:
+			LOG_DBG("symtab at %d", i);
+			ldr->sects[LLEXT_SECT_SYMTAB] = shdr;
+			ldr->sect_map[i] = LLEXT_SECT_SYMTAB;
+			str_cnt++;
+			break;
+		case SHT_STRTAB:
+			if (ldr->hdr.e_shstrndx == i) {
+				LOG_DBG("shstrtab at %d", i);
+				ldr->sects[LLEXT_SECT_SHSTRTAB] = shdr;
+				ldr->sect_map[i] = LLEXT_SECT_SHSTRTAB;
+			} else {
+				LOG_DBG("strtab at %d", i);
+				ldr->sects[LLEXT_SECT_STRTAB] = shdr;
+				ldr->sect_map[i] = LLEXT_SECT_STRTAB;
+			}
+			str_cnt++;
+			break;
+		default:
+			break;
+		}
+	}
+
+	if (!ldr->sects[LLEXT_SECT_SHSTRTAB].sh_type ||
+	    !ldr->sects[LLEXT_SECT_STRTAB].sh_type ||
+	    !ldr->sects[LLEXT_SECT_SYMTAB].sh_type) {
+		LOG_ERR("Some sections are missing or present multiple times!");
+		return -ENOENT;
+	}
+
+	pos = ldr->hdr.e_shoff;
+
+	/* Copy over useful sections */
+	for (i = 0; i < ldr->hdr.e_shnum; i++) {
+		llext_seek(ldr, pos);
+		llext_read(ldr, &shdr, sizeof(elf_shdr_t));
+
+		pos += ldr->hdr.e_shentsize;
+
+		elf32_word str_idx = shdr.sh_name;
+
+		llext_seek(ldr, ldr->sects[LLEXT_SECT_SHSTRTAB].sh_offset + str_idx);
+		llext_read(ldr, name, sizeof(name));
+		name[sizeof(name) - 1] = '\0';
+
+		LOG_DBG("section %d name %s", i, name);
+
+		enum llext_mem mem_idx;
+		enum llext_section sect_idx;
+
+		if (strncmp(name, ".text", sizeof(name)) == 0) {
+			mem_idx = LLEXT_MEM_TEXT;
+			sect_idx = LLEXT_SECT_TEXT;
+		} else if (strncmp(name, ".data", sizeof(name)) == 0) {
+			mem_idx = LLEXT_MEM_DATA;
+			sect_idx = LLEXT_SECT_DATA;
+		} else if (strncmp(name, ".rodata", sizeof(name)) == 0) {
+			mem_idx = LLEXT_MEM_RODATA;
+			sect_idx = LLEXT_SECT_RODATA;
+		} else if (strncmp(name, ".bss", sizeof(name)) == 0) {
+			mem_idx = LLEXT_MEM_BSS;
+			sect_idx = LLEXT_SECT_BSS;
+		} else {
+			LOG_DBG("Not copied section %s", name);
+			continue;
+		}
+
+		ldr->sects[sect_idx] = shdr;
+		ldr->sect_map[i] = sect_idx;
+
+		ext->mem[mem_idx] =
+			k_heap_alloc(&llext_heap, ldr->sects[sect_idx].sh_size, K_NO_WAIT);
+		llext_seek(ldr, ldr->sects[sect_idx].sh_offset);
+		llext_read(ldr, ext->mem[mem_idx], ldr->sects[sect_idx].sh_size);
```

Comment:
Some sections must be 0 initialized and not copied from elf.
You should check that in `shdr.sh_type & SHT_PROGBITS`.

Reply:
Indeed, its likely .bss is incorrect being handled at this time. I don't believe that's necessary to the state initial goal of "hello world" working though, as it assuredly does.

I'm happy to start creating some github issues and a project to track all the wonderful improvements and bug fixes to this initial PoC PR. There's many already in the wild from @pillo79 and @lyakh I believe. I myself have a few branches with changes as well.

My goal with this initial PR is as state though. A starting point for an API, a sample, shell module, and working test showing hello world level functionality in action.